prog/hints.go

```diff
@@ -83,10 +123,182 @@ func generateHints(p *Prog, compMap CompMap, c *Call, arg Arg, exec func(newP *P
 	}
 }
 
-func checkConstArg(arg *ConstArg, compMap CompMap, cb func(newArg Arg)) {
-	for v, _ := range compMap[arg.Val] {
-		cb(constArg(arg.typ, v))
+// Creates hints for a ConstArg and calls cb() for each of the obtained
+// programs.
+func checkConstArg(arg *ConstArg, compMap CompMap, cb func(arg, newArg Arg)) {
+	replacersSet := getReplacersForVal(arg.Val, compMap)
+	for newV, _ := range replacersSet {
+		newArg := constArg(arg.typ, newV)
+		cb(arg, newArg)
+	}
+}
+
+// Returns a set of replacers for a given value. A replacer is a value with
+// which an argument's value should be replaced.
+func getReplacersForVal(value uint64, compMap CompMap) uint64Set {
+	replacersSet := make(uint64Set)
+	f := func(transform func(uint64) uint64) {
+		transformedValue := transform(value)
+		originalsSet := make(uint64Set)
+		// Get all different mutations of transformedValue: shrink, expand, ...
+		vals := getMutationsForConstVal(transformedValue)
+		diff := addArrayToSet(originalsSet, vals)
```

Comment:
originalsSet is empty, so diff == vals, so you can remove originalsSet and addArrayToSet and just use vals.

Reply:
OK, I see what this does. I think it's better to collect values to a map right away.